beacon_node/store/src/hot_cold_store.rs

```diff
@@ -1418,6 +1424,93 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
             &CompactionTimestamp(compaction_timestamp.as_secs()),
         )
     }
+
+    /// Try to prune all execution payloads, returning early if there is no need to prune.
+    pub fn try_prune_execution_payloads(&self, force: bool) -> Result<(), Error> {
+        let split = self.get_split_info();
+
+        if split.slot == 0 {
+            return Ok(());
+        }
+
+        let bellatrix_fork_slot = if let Some(epoch) = self.spec.bellatrix_fork_epoch {
+            epoch.start_slot(E::slots_per_epoch())
+        } else {
+            return Ok(());
+        };
+
+        // Load the split state so we can backtrack to find execution payloads.
+        let split_state = self.get_state(&split.state_root, Some(split.slot))?.ok_or(
+            HotColdDBError::MissingSplitState(split.state_root, split.slot),
+        )?;
+
+        // The finalized block may or may not have its execution payload stored, depending on
+        // whether it was at a skipped slot. However for a fully pruned database its parent
+        // should *always* have been pruned.
+        let split_parent_block_root = split_state.get_block_root(split.slot - 1)?;
+        if !self.execution_payload_exists(&split_parent_block_root)? && !force {
+            info!(self.log, "Execution payloads are pruned");
+            return Ok(());
+        }
+
+        // Iterate block roots backwards to the Bellatrix fork or the anchor slot, whichever comes
+        // first.
```

Comment:
Rough proof that this is correct:

- If the node was checkpoint synced before Bellatrix, then it will have full states all the way back to the Bellatrix fork and will be capable of iterating block roots back to the fork boundary. If we implement pruning of finalized states in future (which I've started on for https://github.com/sigp/lighthouse/pull/3511) then we will need to tell users to ensure their DBs are payload-pruned _before_ enabling state pruning, or we will need to update the iterator to use the linear array of block roots from the freezer DB.
- If the node was checkpoint synced after Bellatrix then it will have applied full blocks forwards from the `anchor_slot`. It _cannot_ have any execution payloads stored prior to the anchor slot because the blocks prior to the anchor are filled in via backfill which stores blinded blocks. Further, nodes are always able to iterate block roots back to their anchor slot because they have a full state stored at least every 8192 slots (max SPRP), so the anchor slot will be in-range of some frozen state. This relies on the same assumption as above about the absence of state pruning.

Reply:
> Further, nodes are always able to iterate block roots back to their anchor slot because they have a full state stored at least every 8192 slots (max SPRP), so the anchor slot will be in-range of some frozen state.

Say we checkpoint sync at slot 32,768 (8192 * 4). Then, we are 50% through the back-fill which means our `anchor_slot` is 16,384 (8192 * 2). I don't see how we have a full state stored at 49152 (8192 * 3) which allows us to iter back to the `anchor_slot`.

Reply:
The anchor slot is the slot of the checkpoint state, not the slot of the oldest backfilled block (that's `oldest_block_slot`)

Reply:
Sorry my proof was a bit unclear: there will be freezer states at least every 8192 slots _after the anchor slot_.

Reply:
> The anchor slot is the slot of the checkpoint state, not the slot of the oldest backfilled block (that's oldest_block_slot)

Oh sorry, of course. I got mixed up there :pray: